### PR TITLE
Make all buffers (actions, rewards, truncation, termination) same length as observation buffer

### PR DESCRIPTION
**Pull request by tomekster**
# Description
Change the behaviour of DataCollector by adding a dummy action, reward, truncation and termination entry at the beginning of every new episode.

This is to keep the size of all buffers the same as the size of the observations buffer (which previously included one more initial observation)

## Type of change

> Please delete options that are not relevant.

- New feature (non-breaking change which adds functionality)
- Breaking change (fix or feature that would cause existing functionality to not work as expected)
- This change requires a documentation update

### Screenshots

# Checklist:

- [x] I have run the [`pre-commit` checks](https://pre-commit.com/) with `pre-commit run --all-files` (see `CONTRIBUTING.md` instructions to set it up)
- [x] I have run `pytest -v` and no errors are present.
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] I solved any possible warnings that `pytest -v` has generated that are related to my code to the best of my knowledge.
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes

### CODE REVIEW

**Review by younik — CHANGES_REQUESTED** [review]
Overall looks good, thank you!

I added some comments on the structure; you should then fix tests and add a test checking this functionality (basically checking that in any case the lengths of the EpisodeData values are all the same)

**Review by younik — CHANGES_REQUESTED** [review]
Overall looks good, I think it is almost ready!

We also need to update the documentation and make the changes very clear, like here: https://minari.farama.org/content/dataset_standards/

File: minari/data_collector/data_collector.py [hunk 1]
```diff
@@ -275,7 +311,8 @@ def reset(
 
     def _validate_buffer(self):
         if len(self._buffer) > 0:
-            if "actions" not in self._buffer[-1].keys():
+            rewards = self._buffer[-1]["rewards"]
+            if len(rewards) == 1 and np.isnan(rewards[0]):
```
[the remark this answers is not in the document]
Comment (tomekster) [reply minari/data_collector/data_collector.py:315]: Yes, that's correct, I'll remove the first condition

File: tests/common.py [hunk 2]
```diff
@@ -559,26 +559,29 @@ def check_data_integrity(data: MinariStorage, episode_indices: Iterable[int]):
 
     # verify the actions and observations are in the appropriate action space and observation space, and that the episode lengths are correct
     for episode in episodes:
+        offset = 1 if np.isnan(episode["rewards"][0]) else 0
```
[the remark this answers is not in the document]
Comment (tomekster) [reply tests/common.py:562]: This is going to break the test_dtaset_download.py tests. We could either:
a) Regenerate those datasets and upload them to Farama server
b) Remove those tests
c) Leave the conditional offset to support the old format

You are leaning towards a) if I understand it correctly and want to use solely the new format moving forward?

[the remark this answers is not in the document]
Comment (younik) [reply tests/common.py:563]: Yes, I was thinking about regenereting the dataset (as we will do it anyway for arrow), but thiking more carefully, I think we should keep the same number of total_steps as before, thus offset must remain 0. After all, we didn't change the number of steps

**Comment by jamartinh** [comment]
Hi, it seems that this is at pull request stage, but I don't agree with the intention.
Indeed the idea to add the rested observation at index 0 is a bad idea, the observations at index 0 and index 1 are the same.
So there is not info loss by not grabbing these observation since it is the first.
The only remaining thing is the first info which can be added to info's as a field called "info/firstinfo"

**Comment by younik** [comment]
> the observations at index 0 and index 1 are the same

This is not true, the observation at index 0 is the observation returned by `env.reset()`, while at index 1 there is the first observation of `env.step`, which is different (it is the result of taking the first action). 

This PR aims to pad actions, rewards, terminations, and truncations to have the episode buffer with equal lengths. This ease storing buffers in a tabular form and in contiguous tensors.

**Comment by jamartinh** [comment]
@younik  Yes, it is incorrect.

Currently, what I do to have an episode to train is:

```python

ep_dict = {
                    "id": ep_group.attrs.get("id"),
                    "total_timesteps": ep_group.attrs.get("total_steps"),
                    "seed": seed,
                    "observations": ep_group["observations"][:][:-1],
                    "next_observations": ep_group["observations"][:][1:],
                    "actions": ep_group["actions"][:],
            }
```

Why is the reason to insert dummy things? 
If you insert a dummy action and terminated and truncated at index 0 you are misaligning the data from the standard (s,a,r,s',t) line. don't you think?

**Comment by younik** [comment]
I don't think I understood your first message then. In your example, `ep_group["observations"][0]` is different from `ep_group["observations"][1]`. Moreover, we are not modifying observations, we are padding the actions.

The advantage is for storing the data in a tabular format. I agree that this may be confusing, I am still discussing how this should be done, so I appreciate your feedback.

**Comment by jamartinh** [comment]
Hi @younik , I understand that from the implementation POV it makes sense and it is desirable to have this feature, however this issue has appeared now many times even from some years ago for the d3rlpy library with its first version of MDPDataset that at first was misaligned with (s,a,r,s',a',t) and was a source of much confusion for users, then at the begining of this project (Minari) it was the same, there wa a question by me some months ago about this precisely, the alignment, and now we are again with this.

Any practitioner of RL is expecting that if he see a tuple of ( state (s), action (a) , reward (r) , next_state (s'),   done (t)) this tuple is "aligned", meaning that the action (a)  is taken at state (s) and reward  (r) is obtained as a result of env.step.

This is not only in the mind, but in many implementations such as ReplayBuffers etc.
So, Having a stored format with this misalignment makes the physical file (now the hdf5 file) able to lead to future silent errors if new developers/users either extend it or use directly the datasets (hdf5 files) without using minari layer as an "interpreter".

Perhaps, instead of adding dummies at index 0 it is better at last index so "appending" instead of inserting (padding) at index 0.

IMHO, we have to try either enforcing an explicit error when not using it well, such as "bad index" raised by python itself or a silent error that does not damage learning.

For instance, will you repeat the truncated and terminated values with a value of 1 ? of put it to 0 ?
If you don't repeat truncated or terminated some users after concatenating all the episodes in a table may commit the mistake of doing some calculation assuming that the episode limits are the indexes of a numpy array or tensor when terminated or truncated have value of 1.

Issues has arised with D4RL datasets regarding this, and some tricks where used such as checking if the observations of the row n+1 != next_observation of row n. So users used that for checking episode boundaries.

So, in esense, what I ask is to threat this alignment and episode boundaries as a very important thing, after all the central part of a library as minari is the definition of a standard dataset format (as I understand) and a set of easy tools to generate and exploit them.

Cheers,
JAMH

**Comment by younik** [comment]
> Perhaps, instead of adding dummies at index 0 it is better at last index so "appending" instead of inserting (padding) at index 0.

Yes, this was the idea (we were still iterating on this PR).

Anyway, we agree that this is confusing for the user and, after discussing further, we will keep the things as they are. 
While we will need padding anyway for tabular format like Arrow, we will make this hidden to the user (slicing during loading), and always return the data in the current format. 

Thank you for rasing your concerns @jamartinh!

**Pull request closed without merging**